proposals/0001-kubernetes-native-api/README.md

```diff
@@ -0,0 +1,715 @@
+restrictions would apply to both APIs (e.g. pipeline versions are immutable). This means that existing integrations with
+the REST API (e.g. the Kubeflow dashboard) will work seamlessly.
+
+It is the desire to eventually support pipeline runs through a Kubernetes native API, however, this is out of scope for
+this feature as there are scale concerns with etcd (e.g. 80k+ runs by some users) and require a different approach with
+more tradeoffs.
+
+## Motivation
+
+Users have established GitOps practices with tooling such as Argo CD for Kubernetes configuration management and
+application deployments. Allowing users to leverage those tools for Kubeflow pipeline definitions will reduce the
+friction felt when adopting the project and provide a more consistent experience across Kubeflow. It also allows for a
+familiar and simple query experience from the `kubectl` CLI without needing to maintain a KFP CLI.
+
+### Goals
+
+1. A Kubeflow pipelines user can leverage established GitOps practices to deploy pipelines and pipeline versions through
+   a Kubernetes native API.
+1. Existing integrations with the REST API (e.g. Kubeflow dashboard) are not impacted.
+1. The Kubernetes native API has the same validation and restrictions as the existing REST API (e.g. pipeline versions
+   are immutable).
+1. The user can opt-in to the Kubernetes native APIs and is not forced to use them or migrate their existing pipeline
+   definitions to the Kubernetes native API feature.
+1. Provide a sample migration script for users wanting to migrate their existing pipelines and pipeline versions to the
+   Kubernetes native API approach.
+1. The Kubeflow Pipelines Python SDK is enhanced to allow the output format to be a Kubernetes native API manifest.
+
+### Non-Goals
+
+1. Kubeflow pipeline runs will not be included in the Kubernetes native API at this time. It is a desire to eventually
+   do this, but it has scale concerns (e.g. 80k+ runs by some users) that require a different approach with tradeoffs.
+1. This will not support the v1 API of Kubeflow pipelines. To continue using the v1 API, the user must continue using
+   the default database solution.
+
+## Proposal
+
+The proposal is to allow the user to opt-in to a backwards compatible Kubernetes native API, achieved through custom
+resource definitions (CRDs), to define pipelines and pipeline versions. The proposed approach is to use the Kubernetes
+API as the pipeline storage interface in the REST API to make things seamless and backwards compatible while not
+introducing data duplication or synchronization issues.
+
+The same validation and restrictions would apply to both the Kubernetes API and the REST API such as pipeline versions
+being immutable, needing to reference an existing pipeline, and have a valid pipeline spec. This will be achieved
+through the API server having endpoints for Kubernetes validating and mutating webhooks that leverage the same Go code
```

Comment:
We'll have to ensure that the failure modes are the same between the database and CRD backed approach.